File: leetcode/160.intersection-of-two-linked-lists/cpp/getintersectionnode.cpp
#include <unordered_set>

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
    {
        std::unordered_set<ListNode *> nodesInFirstList;

        while (headA != nullptr)
        {
            nodesInFirstList.emplace(headA);
            headA = headA->next;
        }

        while (headB != nullptr)
        {
            if (nodesInFirstList.find(headB) != nodesInFirstList.end())
                return headB;
            headB = headB->next;
        }

        return nullptr;
    }
};

File: leetcode/160.intersection-of-two-linked-lists/go/getintersectionnode.go
package getintersectionnode

type ListNode struct {
	Val  int
	Next *ListNode
}

// M+N, M
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	nodesInFirstList := make(map[*ListNode]struct{})

	for headA != nil {
		nodesInFirstList[headA] = struct{}{}
		headA = headA.Next
	}

	for headB != nil {
		if _, exist := nodesInFirstList[headB]; exist {
			return headB
		}
		headB = headB.Next
	}

	return nil
}
